### PR TITLE
Run MLprojects on kubernetes

### PR DESCRIPTION
**Pull request by marcusrehm**
## What changes are proposed in this pull request?
 
This PR proposes run project execution on Kubernetes.
  
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
This PR adds a new mode of project execution on kubernetes.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [x] CLI 
- [ ] API 
- [ ] REST-API 
- [x] Examples 
- [x] Docs
- [ ] Tracking
- [x] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by marcusrehm** [comment]
It still lacks documentation and some tests

**Comment by marcusrehm** [comment]
MLproject file in Docker example updated to also run in Kubernetes mode. The docs were also updated.

**Comment by mateiz** [comment]
This is great, Marcus. I'll try find someone with more Kubernetes background to review it in detail.

**Comment by marcusrehm** [comment]
Hi @mateiz, thanks for replying! 
No worries, I'll be reviewing and improving the code while you find someone. 
When I delivered the feature for running MLprojects in Docker containers @smurching gave me a lot of help reviewing it.

File: examples/docker/MLproject
Comment (phatlast96) [inline examples/docker/MLproject]: Can you give another name?

File: mlflow/projects/kubernetes.py
Comment (talgo10) [inline mlflow/projects/kubernetes.py]: What about exposing the **resources.requests** and **resources.limits** as optional flags to the cli command? The user will expect to have control on this. 

Correct me if I'm wrong, the default values are based on the following order:
1. the namespace defaults - https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/
2. kubernetes defaults - https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits

**Comment by andrewmchen** [comment]
Thanks for contributing this! I'm super excited about this feature :)

At a high level this PR looks pretty good. Some feedback on the API:

In your PR, we put the `job_namespace`, `image_namespace`, and `registry` arguments in the MLProject spec. The problem with this is that MLProject specs should define a project which is potentially re-usable by many different users. For example, we published some example ones here: https://databricks.com/blog/2018/08/16/introducing-mlflow-apps-a-repository-of-sample-applications-for-mlflow.html. Instead all of these arguments should probably go in as arguments to `mlflow.run`.

The other main feedback I have is that there are a ton of different possible k8s configs we can imagine a user providing. @talgo10 already suggested two reasonable ones but perhaps a user wants to provide Volumes, Secrets, future K8S features, w/e :). Could we configure these by allowing the user to pass in an optional yaml string which defines their K8S `batch/V1Job` object? When we actually run the job, we can merge in the required fields like the image URI and command and such.

File: examples/docker/MLproject
Comment (andrewmchen) [inline examples/docker/MLproject]: Hm, is there something special about the URI structure `registry/namespace/name`? Would it be more simple to ask the user to provide just an `image_name_prefix` or something?

File: mlflow/cli.py
Comment (andrewmchen) [inline mlflow/cli.py]: Is this necessary? Doesn't the docker python client use your local docker daemon's authentication information already?

Comment (marcusrehm) [reply mlflow/cli.py]: I think it's necessary because sometimes we may need to push to different repositories. As you said above, users may have different needs, like different registries per team. So thinking about these use cases, we need to pass credentials.
What do you think?

File: mlflow/projects/kubernetes.py
Comment (marcusrehm) [reply mlflow/projects/kubernetes.py]: hey @talgo10, that's a great idea. Actually I was working on a way to let users pass a path to a job template file. I've just pushed it, but I created the resources attribute in MLproject file too because I think this is an attribute that will be frequently used as you mentioned.

File: examples/docker/MLproject
Comment (marcusrehm) [reply examples/docker/MLproject]: hey @andrewmchen , thanks for your comments!
Actually, as I said above to @talgo10, now it's possible to pass a template file for the job. There is a simple one inside the docker example folder.

Comment (marcusrehm) [reply examples/docker/MLproject]: About the `job_namespace`, it's used to specify in which k8s namespace the job show be deployed. I think the job namespace where the training will be executed would be more of a project definition than a users choice, that's why I let it in the MLproject file.

File: mlflow/cli.py
Comment (marcusrehm) [reply mlflow/cli.py]: Complementing, as I said above, regarding the `job_namespace`, I think the same applies to publish images. So the registry and image namespace would be in MLproject file.

File: examples/docker/MLproject
Comment (marcusrehm) [reply examples/docker/MLproject]: Complementing, as I said above, regarding the `job_namespace`, I think the same applies to publish images. So the registry and image namespace would be in MLproject file.

File: mlflow/cli.py
Comment (andrewmchen) [reply mlflow/cli.py]: Would you need this functionality yourself? I feel the common case is that the docker daemon's authentication information is enough for most use cases. As such, I think it's better to widen the API when more concrete use cases come up in the future.

File: examples/docker/MLproject
Comment (andrewmchen) [inline examples/docker/MLproject]: Mentioned this briefly above but going to repeat just in case. We shouldn't put things like `job_namespace`, `image_namespace` and `registry` or other such arguments here which are not generalizable to all users of this MLProject.

File: examples/docker/MLproject
Comment (andrewmchen) [reply examples/docker/MLproject]: `job_namespace` is not generalizable to all users right? If you say this project should be run in `namespace = marcusrehm` and I don't have this namespace in my k8s cluster, the project run will fail.

Same with `registry` and `image_namespace`: if you choose a registry I don't have push access to my run will fail.

Comment (marcusrehm) [reply examples/docker/MLproject]: hey @andrewmchen ! I got your point. I was thinking about our process here (which led me to develop the feature) and your approach is better and covers more use cases.

Comment (andrewmchen) [reply examples/docker/MLproject]: Awesome, BTW I just wanted to make sure you knew about this refactoring that @aarondav  is doing to get ready for 1.0. https://github.com/mlflow/mlflow/pull/1222/files#diff-67cc4fcc56ad80e1028ac5d959b05985R71

It might make sense to put the `k8s_context`, `k8s_job_template`, and `docker_registry` fields in that file as well.

File: examples/docker/MLproject
Comment (marcusrehm) [reply examples/docker/MLproject]: I see, so right now we would have just the resources stuff, related to pods quotas and limits. I think that we could then, remove the `kubernetes_env` section and let everything in the backend_config.
Do you agree?

File: examples/docker/MLproject
Comment (marcusrehm) [reply examples/docker/MLproject]: I did the update in my branch. I'm still doing the changes but right now this commit is already merged. :)

File: examples/docker/MLproject
Comment (marcusrehm) [reply examples/docker/MLproject]: Hi @andrewmchen ! 

I finished all the fixes and improvements in order to use this feature with the refactoring that @sueann did.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: Just curious, why is this different than the image in the `MLProject` file?

File: mlflow/projects/_project_spec.py
Comment (andrewmchen) [inline mlflow/projects/_project_spec.py]: Do we still use `kubernetes_env`, if not let's remove all usages of it.

File: mlflow/cli.py
Comment (andrewmchen) [inline mlflow/cli.py]: Could we refactor this line to say `Invalid backend config JSON. Parse error...`?

File: mlflow/cli.py
Comment (andrewmchen) [inline mlflow/cli.py]: Could we refactor all these `cluster_spec_arg`s to maybe just `backend_config`?

File: mlflow/projects/_project_spec.py
Comment (marcusrehm) [reply mlflow/projects/_project_spec.py]: Sorry, we are not using it anymore.

Comment (marcusrehm) [reply mlflow/projects/_project_spec.py]: done!

File: mlflow/cli.py
Comment (marcusrehm) [reply mlflow/cli.py]: done!

File: mlflow/cli.py
Comment (marcusrehm) [reply mlflow/cli.py]: done!

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: As docker refers to environment and kubernetes to a backend mode, I thought would be better to let kubernetes config apart from the docker env. Now projects don't need to have a docker_env section if it will run on kubernetes, just the backend_config file.

Comment (andrewmchen) [reply mlflow/projects/__init__.py]: I think this distinction actually make the user workflows more complicated. For example, what happens when I run a MLproject with a `conda_env`?

Instead there are two dimensions here: one is what the project looks like (conda based, docker based), and one is how the project is executed (locally, databricks, k8s). In the future, we can imagine running these docker projects on databricks as well, in that case we don't want to redefine the docker base_image flag for that execution backend for example.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: What's the behavior if the `synchronous` flag is equal to false. If it's not actually async we should throw an exception here.

Comment (marcusrehm) [reply mlflow/projects/__init__.py]: It's the same behavior as `mode == databricks`, it ignores the flag and runs synchronous. AFAIK this flag is only used in local mode, isn't it?

Comment (marcusrehm) [reply mlflow/projects/__init__.py]: The point about throw an exception is that today the sychronous parameter gets it's value from cli.py in the block below:
```
    try:
        projects.run(
            uri,
            entry_point,
            version,
            experiment_name=experiment_name,
            experiment_id=experiment_id,
            parameters=param_dict,
            backend=backend,
            backend_config=backend_config,
            use_conda=(not no_conda),
            storage_dir=storage_dir,
            synchronous=backend == "local" or backend is None,
            run_id=run_id
        )
    except projects.ExecutionException as e:
        _logger.error("=== %s ===", e)
        sys.exit(1)
```

So this will be always false for `backends` different than `local`. I think that we should let it as it is now. What do you think?

Comment (andrewmchen) [reply mlflow/projects/__init__.py]: I see. I overlooked the fact that it is not used if `mode=databricks`. Makes sense to leave it for now then.

BTW, did you see this comment yet? https://github.com/mlflow/mlflow/pull/1181#discussion_r286227289

Comment (marcusrehm) [reply mlflow/projects/__init__.py]: Hey @andrewmchen ! As we spoke earlier, I changed the implementation to use the `docker_env` again.

Comment (marcusrehm) [reply mlflow/projects/__init__.py]: Just need to update the docs

Comment (marcusrehm) [reply mlflow/projects/__init__.py]: done @andrewmchen ! I think it's ready now :)

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: I think we need to return a `SubmittedRun` object here which implements a wait function which actually implements the polling for the completion of the k8s project.

For example: https://github.com/mlflow/mlflow/blob/master/mlflow/projects/databricks.py#L325.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: I think the project env should be `docker` right? Maybe we need a new tag called project execution mode.

File: docs/source/projects.rst
Comment (andrewmchen) [inline docs/source/projects.rst]: Is this a breaking change or is it just a problem with the docs. It seems like before this PR it was already looking at the `image` field.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: Is this a breaking change from before this PR?

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: Will this line ever run? Seems like if `backend_config` is none, the previous line will throw.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: I don't think the `kube-job-template-path` should be relative to `MLproject` since it may contain user specific configurations like namespace.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: might make sense to use `_validate_docker_env` here as well.

File: mlflow/projects/kubernetes.py
Comment (andrewmchen) [inline mlflow/projects/kubernetes.py]: It looks like this line doesn't always throw on error cases. See the following:
```
ipdb> p client.images.push(repository=image_uri)
u'{"status":"The push refers to repository [mlflow-registry.dev.databricks.com/andrew/mlflow-test]"}\r\n{"errorDetail":{"message":"Get https://mlflow-registry.dev.databricks.com/v2/: Service Unavailable"},"error":"Get https://mlflow-registry.dev.databricks.com/v2/: Service Unavailable"}\r\n'
ipdb> p type(client.images.push(repository=image_uri))
<type 'unicode'>```

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: If I set my `image-uri` to something like `"registry.dev.databricks.com/andrew/mlflow-test"` this won't work since it changes the host to `mlflow-registry.dev.databricks.com`. How are you using `image-uri` to get it to work?

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: I see, but the point here is that the monitor is reading a stream from pod's log. It's not polling an REST API as we do in Databricks mode. We can think about move this logic to a `SubmittedRun` object though.

What do you think?

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: Yes you're right. I've created a tag called `backendMode`, so we can have it with the same name as parameter okay?

File: docs/source/projects.rst
Comment (marcusrehm) [reply docs/source/projects.rst]: Yes, the docs were wrong.

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: Sorry, my mistake, I should have ran it here. This issue was introduced in another PR, the [PR 1331](https://github.com/mlflow/mlflow/pull/1331), I've just pulled it so this branch could be able to merge to master.

Done.

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: `os.path.join` just joins the two paths only if the second is relative (without a `/` in the beginning), so if you pass the absolute path it won't join with the mlflow work_dir.

You can check here in [docs](https://docs.python.org/2/library/os.path.html#os.path.join):
> If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.

I think we can let as it is so both use cases, absolute and relative paths, are valid. What do you think?

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: Yes, now that we are using it again, it makes sense.

Done.

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: you're right, I've changed it.

File: mlflow/projects/kubernetes.py
Comment (marcusrehm) [reply mlflow/projects/kubernetes.py]: yes, this one was tricky. The error isn't an API error, but now it's checking for errors in response.

Done

File: mlflow/projects/__init__.py
Comment (andrewmchen) [reply mlflow/projects/__init__.py]: It's similar, when the function `run` invokes `_run`, it expects for `_run` to return a `SubmittedRun` object which it can wait/poll for completion on. The nice thing about this interface is now we can run mlflow jobs asynchronously by simply not blocking for the `SubmittedRun`'s completion.

File: mlflow/projects/__init__.py
Comment (andrewmchen) [reply mlflow/projects/__init__.py]: Since the `backend_config` is loaded relative to the CWD, I think it only makes sense that the `kube-job-template-path` is also loaded relative to the CWD. We can discuss more offline if you'd like!

File: mlflow/projects/__init__.py
Comment (andrewmchen) [inline mlflow/projects/__init__.py]: I guess this is a breaking change for docker projects? Can we put back the mlflow prefix for docker projects only?

File: mlflow/projects/kubernetes.py
Comment (andrewmchen) [inline mlflow/projects/kubernetes.py]: Could we provide these through the kube-job-template for now?

File: setup.py
Comment (andrewmchen) [inline setup.py]: Can we remove these from the install_requires? Not everyone who uses mlflow will use k8s projects so we shouldn't force these deps on them.

File: docs/source/projects.rst
Comment (ahutterTA) [inline docs/source/projects.rst]: nit: s/follwing/following

File: setup.py
Comment (marcusrehm) [reply setup.py]: I see. Okay. Removed.

File: docs/source/projects.rst
Comment (marcusrehm) [reply docs/source/projects.rst]: Done

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: It doesn't break anything, I've tested here and everything is running fine. Actually, when I first developed the docker feature it didn't need that too.

File: mlflow/projects/kubernetes.py
Comment (marcusrehm) [reply mlflow/projects/kubernetes.py]: Sure, I removed it.

We can let users provide it in base image so they can build it with these variables values. Thinking about security this is a better choice because these keys wouldn't be exposed in config files. What do you think?

File: mlflow/projects/__init__.py
Comment (marcusrehm) [reply mlflow/projects/__init__.py]: I've just checked it here, the loading of  `backend_config` also behave this way, We can pass an absolute or relative path and the `open` function will handle the same way `os.path.join` handles it. You can see the [docs](https://docs.python.org/3/library/functions.html#open) here.
> file is a path-like object giving the pathname (absolute or relative to the current working directory) 

We can still chat offline if you'd like too.

**Review by tomasatdatabricks — APPROVED** [review]
Looks good! Thanks for the contribution @marcusrehm!

Fyi, I have updated the logic for monitoring job status (pushed to your branch). I think the previous version was using indirect way (active, failed, succeeded counts) of figuring out when the job is done and when it is successful. Someone else should review that part after me. 

Other than that, lgtm!

File: mlflow/cli.py
Comment (andrewmchen) [inline mlflow/cli.py]: Should we make it async by default?

File: travis/lint-requirements.txt
Comment (andrewmchen) [inline travis/lint-requirements.txt]: Should we revert this diff?

File: mlflow/utils/mlflow_tags.py
Comment (andrewmchen) [inline mlflow/utils/mlflow_tags.py]: Should this be just `mlflow.project.backend`?

File: docs/source/projects.rst
Comment (andrewmchen) [inline docs/source/projects.rst]: spec.template.spec.containers[0].name

File: docs/source/projects.rst
Comment (andrewmchen) [inline docs/source/projects.rst]: we should include a `backoffLimit: 0` here by default.

File: mlflow/cli.py
Comment (marcusrehm) [reply mlflow/cli.py]: I think it would be better this way, so we can monitoring the execution easier. The other way would be to grab info from kubernetes with kubectl or monitor mlflow ui to see when the run had finished. What do you think?

File: mlflow/utils/mlflow_tags.py
Comment (marcusrehm) [reply mlflow/utils/mlflow_tags.py]: yes, we can change it.

**Comment by marcusrehm** [comment]
@tomasatdatabricks thanks for your help! I can review the logic for monitoring job status.

File: docs/source/projects.rst
Comment (marcusrehm) [reply docs/source/projects.rst]: Done.

File: docs/source/projects.rst
Comment (marcusrehm) [reply docs/source/projects.rst]: Done.

**Comment by tomasatdatabricks** [comment]
@marcusrehm Awesome, thanks!

File: mlflow/utils/mlflow_tags.py
Comment (marcusrehm) [reply mlflow/utils/mlflow_tags.py]: Done

File: mlflow/cli.py
Comment (marcusrehm) [reply mlflow/cli.py]: @tomasatdatabricks and @andrewmchen , what do you guys think about get pod's log in case of errors? I'm reviewing @tomasatdatabricks 's changes for job monitoring and I'm getting some errors from pods execution. Now I need to run `kubectl` to find the pod related to the job and get it's logs.

We could ease the debugging if we just print out the logs in case of errors. We can chat on Slack if you like.

File: travis/lint-requirements.txt
Comment (marcusrehm) [reply travis/lint-requirements.txt]: Yes, I was getting some errors in the beginning while trying to run `lint.sh`, but now I tested without it and it's okay.

Comment (marcusrehm) [reply travis/lint-requirements.txt]: yes, it is required @andrewmchen . Please see: https://travis-ci.org/mlflow/mlflow/jobs/549827401

File: mlflow/cli.py
Comment (tomasatdatabricks) [reply mlflow/cli.py]: Hey @marcusrehm. 

That is a good idea. But I think the current PR provides good functionality as is and we can leave this for a follow up PR.

It's not 100% clear what is the right thing to do regarding the logs. For example, we might also pull them after every run and attach them to the run as artifact(s).  Streaming the output as you had it in your PR may also be a good idea. It may be confusing thought in case multiple pods run during the execution (either due to parallelism or retry).  .

Comment (marcusrehm) [reply mlflow/cli.py]: Hey @tomasatdatabricks , I think you're right. We can let it for another PR. I've tested the logic you changed and it seams okay!

If you and @andrewmchen think we have nothing left to change then I think we can close the PR.

Comment (marcusrehm) [reply mlflow/cli.py]: The build is broken but it's related to spark testing, not to what we are doing.

Comment (tomasatdatabricks) [reply mlflow/cli.py]: I'll take a look. Maybe we just need to rerun.

Comment (marcusrehm) [reply mlflow/cli.py]: Thanks @tomasatdatabricks !

**Pull request merged**
